[GISel] Teach computeKnownBitsImpl to handle COPY instructions that change bit width.

topperc opened the pull request:
The sexti32 ComplexRenderFn on RISCV calls computeNumSignBits which calls computeKnownBits.

I encountered a case where we looked through a G_PHI and found a COPY that was created from an already selected G_TRUNC from s64 to s32. s32 and s64 integers on RISC-V end up in the same register class. s32 G_PHI is legal to allow f32 phis on RV64.

This patch uses KnownBits::anyextOrTrunc to adjust the width when they mismatch.

topperc committed:
[GISel] Teach computeKnownBitsImpl to handle COPY instructions that change bit width.

The selectShiftMask ComplexRenderFn on RISCV calls computeKnownBits.

I encountered a case where we looked through a G_PHI and found a
COPY that was created from an already selected G_ANYEXT from s32 to
s64. s32 and s64 integers on RISC-V end up in the same register class.a
The input to the COPY was an already selected s32 SELECT instruction.
We need an s32 SELECT to be legal to support f32 selects. If it isn't
used by FP operations, regbank select will assign to GPR.

This patch uses KnownBits::anyextOrTrunc to adjust the width when
they mismatch.

I haven't reduced a test case yet, but wanted to make sure this is
the right fix.

arsenm commented:
> I haven't reduced a test case yet, but wanted to make sure this is the right fix.

Easier to judge this with the test but probably

topperc commented:
While investigating this, I realized the we shouldn't be promoting s32 G_PHI to s64 on RV64 or we can't support phis of f32 values. Fixing that made the problem disappear for now.

topperc commented:
I've added a test now.

topperc committed:
fixup! Add -o - to test.

The update script doesn't require this to generate the test, but
running the test through lit does.
